keps/sig-node/3386-kubelet-evented-pleg/README.md

```diff
@@ -134,15 +134,62 @@ This might be a good place to talk about core concepts and how they relate.
+This feature can only be used when `EventedPLEG` feature gate is enabled.
+
+### Timestamp of the Pod Status
+![Existing Generic PLEG](./existing-generic-pleg.png)
+
+Kubelet cache saves the [pod status with the timestamp](https://github.com/kubernetes/kubernetes/blob/c012d901d8bee86ef3e3c9472a1a4a0368a34775/pkg/kubelet/pleg/generic.go#L426). The value of this timestamp is calculated [within the kubelet process](https://github.com/kubernetes/kubernetes/blob/c012d901d8bee86ef3e3c9472a1a4a0368a34775/pkg/kubelet/pleg/generic.go#L399). This works fine when there is only Generic PLEG at work as it will calculate the timestamp first and then fetch the `PodStatus` to save it in the cache.
+
+As of today, the `PodStatus` is saved in the cache without any validation of the existing status against the current timestamp. This works well when there is only `Generic PLEG` setting the `PodStatus` in the cache.
+
+If we have multiple entities, such as `Evented PLEG`, while trying to set the `PodStatus` in the cache we may run into the racy timestamps given each of them were to calculate the timestamps in their respective execution flow. While `Generic PLEG` calculates this timestamp and gets the `PodStatus`, we can only calculate the corresponding timestamp in `Evented PLEG` after the event has been received by the Kubelet. Any disruptions in getting the events, such as errors in the grpc connection, might skew our calculation of the time in the kubelet for the `Evented PLEG`.
+
+In order to address the issues above, we propose that existing `Generic PLEG` as well as `Evented PLEG` should rely on the CRI Runtime for the timestamp of the `PodStatus`. This way the `PodStatus` would also be a bit more closer to the actual time when the statuses of the `Sandboxes` and `Containers` where provided by the CRI Runtime. It will enable us to correctly compare the timestamps before saving them in the cache, to avoid the erroneous behaviour. This should also prevent any old buffered `PodStatus` (consolidated during any disruptions or failures) from overriding the newer entry in the cache.
+
+![Modified Generic PLEG](./modified-generic-pleg.png "Existing Generic PLEG")
+
+![Evented PLEG](./evented-pleg.png)
+
 
 ### Runtime Service Changes
 
-A new RPC will be introduced in the [CRI Runtime Service](https://github.com/kubernetes/kubernetes/blob/6efd6582df2011f1ec8c146ef711b3348ae07d60/staging/src/k8s.io/cri-api/pkg/apis/runtime/v1/api.proto#L34),
+Instead of getting the `Sandbox` and `Container` statuses independently and using the timestamp calculated from the kubelet process, `Generic PLEG` can fetch the `PodStatus` directly from the CRI Runtime using the new RPC which should also have the embedded timestamp coming from the CRI Runtime.
+
+```protobuf=
+service RuntimeService {
+    </snip>
+    // GetPodStatus retrieves the Pod Status from the CRI Runtime
+    rpc GetPodStatus(GetPodStatusRequest) returns (GetPodStatusResponse) {}
+}
+
+message GetPodStatusRequest {
```

Comment:
Instead of introducing a new CRI call, is there a possibiltiy where we have a `bool includeContainers` in the existing `PodSandboxStatus` call?

Reply:
Yes, we can also embed container statuses optionally in PodSandboxStatusResponse, https://github.com/kubernetes/kubernetes/blob/4a894be926adfe51fd8654dcceef4ece89a4259f/staging/src/k8s.io/cri-api/pkg/apis/runtime/v1/api.proto#L535

I can update the code tomorrow to drop new GetPodStatus and use PodSandboxStatusResponse to optionally carry ContainerStatuses

Reply:
+1 on re-using the existing PodSandboxStatus. Only difference is that when reusing PodSandboxStatus, kubelet will need to figure out the podSandboxIds based on the UID, instead of just passing UID to runtime and letting runtime figure out everything. 

I prefer keeping minimum number of APIs if possible, so +1 on reusing.

Reply:
@ruiwen-zhao Good point. 

@mrunalp in Generic PLEG we have access to PodUID, https://github.com/kubernetes/kubernetes/blob/4a894be926adfe51fd8654dcceef4ece89a4259f/pkg/kubelet/pleg/generic.go#L403, not the sandbox Ids

Reply:
The API that I had introduced, GetPodStatus works with PodUID for that reason.